From 60d6a76dcee2a5647d69874d9b5c24f701a6722d Mon Sep 17 00:00:00 2001
From: Geoffrey Thomas <geofft@ldpreload.com>
Date: Mon, 1 Dec 2025 14:11:43 -0500
Subject: [PATCH] getpath: Find the runtime path in shared and static builds
Forwarded: no

Locate the runtime path in shared and static builds. On Linux, try
/proc/self/maps to obtain an absolute path before falling back to dladdr().

Look up the caller's return address so that a function pointer in the
procedure linkage table cannot identify a different object.
---
 Modules/getpath.c | 62 +++++++++++++++++++++++++++++++++++++++++-----
 1 file changed, 56 insertions(+), 6 deletions(-)

diff --git a/Modules/getpath.c b/Modules/getpath.c
index 1e75993480a..347c21e7387 100644
--- a/Modules/getpath.c
+++ b/Modules/getpath.c
@@ -802,14 +802,19 @@ progname_to_dict(PyObject *dict, const char *key)
 }
 
 
+static void
+fclose_cleanup(FILE **pf) {
+    if (*pf) {
+        fclose(*pf);
+        *pf = NULL;
+    }
+}
+
+
 /* Add the runtime library's path to the dict */
 static int
 library_to_dict(PyObject *dict, const char *key)
 {
-/* macOS framework builds do not link against a libpython dynamic library, but
-   instead link against a macOS Framework. */
-#if defined(Py_ENABLE_SHARED) || defined(WITH_NEXT_FRAMEWORK)
-
 #ifdef MS_WINDOWS
     extern HMODULE PyWin_DLLhModule;
     if (PyWin_DLLhModule) {
@@ -817,12 +822,57 @@ library_to_dict(PyObject *dict, const char *key)
     }
 #endif
 
+    const void *target = __builtin_extract_return_addr(
+        __builtin_return_address(0)
+    );
+
+#ifdef __linux__
+    /* Linux libcs do not reliably report the realpath in dladdr dli_fname and
+     * sometimes return relative paths, especially if the returned object is
+     * the main program itself. However, /proc/self/maps will give absolute
+     * realpaths (from the kernel, for the same reason that /proc/self/exe is
+     * canonical), so try to parse and look it up there.  (dyld seems to
+     * reliably report the canonical path, so doing this matches the behavior
+     * on macOS.) */
+
+    __attribute__((cleanup(fclose_cleanup)))
+    FILE *maps = fopen("/proc/self/maps", "r");
+    if (maps != NULL) {
+        /* See implementation in fs/proc/task_mmu.c for spacing. The pathname
+         * is the last field and has any \n characters escaped, so we can read
+         * until \n. Note that the filename may have " (deleted)" appended;
+         * we don't bother to handle that specially as the only user of this
+         * value calls dirname() anyway.
+         * TODO(geofft): Consider using PROCMAP_QUERY if supported.
+         */
+        uintptr_t low, high;
+        char rest[PATH_MAX + 1];
+        while (fscanf(maps, "%lx-%lx %*s %*s %*s %*s", &low, &high) == 2) {
+            if (fgets(rest, PATH_MAX + 1, maps) == NULL) {
+                break;
+            }
+            if (strlen(rest) >= PATH_MAX) {
+                // If the line is too long our parsing will be out of sync.
+                break;
+            }
+
+            if (low <= (uintptr_t)target && (uintptr_t)target < high) {
+                // Skip past padding spaces in the filename.
+                const char *filename = rest + strspn(rest, " ");
+                if (filename[0] == '/') {
+                    return decode_to_dict(dict, key, filename);
+                }
+                break;
+            }
+        }
+    }
+#endif
+
 #if HAVE_DLADDR
     Dl_info libpython_info;
-    if (dladdr(&Py_Initialize, &libpython_info) && libpython_info.dli_fname) {
+    if (dladdr(target, &libpython_info) && libpython_info.dli_fname) {
         return decode_to_dict(dict, key, libpython_info.dli_fname);
     }
-#endif
 #endif
 
     return PyDict_SetItemString(dict, key, Py_None) == 0;
-- 
2.50.1 (Apple Git-155)
